File: app/src/main/java/wangdaye/com/geometricweather/basic/models/weather/Temperature.java
package wangdaye.com.geometricweather.basic.models.weather;

import android.content.Context;

import androidx.annotation.Nullable;

import java.io.Serializable;

import wangdaye.com.geometricweather.basic.models.options.unit.TemperatureUnit;
import wangdaye.com.geometricweather.settings.SettingsOptionManager;

/**
 * Temperature.
 * default unit : {@link wangdaye.com.geometricweather.basic.models.options.unit.TemperatureUnit#C}
 * */
public class Temperature implements Serializable {

    private final int temperature;
    @Nullable private final Integer realFeelTemperature;
    @Nullable private final Integer realFeelShaderTemperature;
    @Nullable private final Integer apparentTemperature;
    @Nullable private final Integer windChillTemperature;
    @Nullable private final Integer wetBulbTemperature;
    @Nullable private final Integer degreeDayTemperature;

    public Temperature(int temperature,
                       @Nullable Integer realFeelTemperature,
                       @Nullable Integer realFeelShaderTemperature,
                       @Nullable Integer apparentTemperature,
                       @Nullable Integer windChillTemperature,
                       @Nullable Integer wetBulbTemperature,
                       @Nullable Integer degreeDayTemperature) {
        this.temperature = temperature;
        this.realFeelTemperature = realFeelTemperature;
        this.realFeelShaderTemperature = realFeelShaderTemperature;
        this.apparentTemperature = apparentTemperature;
        this.windChillTemperature = windChillTemperature;
        this.wetBulbTemperature = wetBulbTemperature;
        this.degreeDayTemperature = degreeDayTemperature;
    }

    public int getTemperature() {
        return temperature;
    }

    @Nullable
    public Integer getRealFeelTemperature() {
        return realFeelTemperature;
    }

    @Nullable
    public Integer getRealFeelShaderTemperature() {
        return realFeelShaderTemperature;
    }

    @Nullable
    public Integer getApparentTemperature() {
        return apparentTemperature;
    }

    @Nullable
    public Integer getWindChillTemperature() {
        return windChillTemperature;
    }

    @Nullable
    public Integer getWetBulbTemperature() {
        return wetBulbTemperature;
    }

    @Nullable
    public Integer getDegreeDayTemperature() {
        return degreeDayTemperature;
    }

    @Nullable
    public String getTemperature(Context context, TemperatureUnit unit) {
        return getTemperature(context, temperature, unit);
    }

    @Nullable
    public String getShortTemperature(Context context, TemperatureUnit unit) {
        return getShortTemperature(context, temperature, unit);
    }

    @Nullable
    public String getRealFeelTemperature(Context context, TemperatureUnit unit) {
        return getTemperature(context, realFeelTemperature, unit);
    }

    @Nullable
    public String getShortRealFeeTemperature(Context context, TemperatureUnit unit) {
        return getShortTemperature(context, realFeelTemperature, unit);
    }

    @Nullable
    public static String getTemperature(Context context, @Nullable Integer temperature, TemperatureUnit unit) {
        if (temperature == null) {
            return null;
        }
        return unit.getTemperatureText(context, temperature);
    }

    @Nullable
    public static String getShortTemperature(Context context,
                                             @Nullable Integer temperature, TemperatureUnit unit) {
        if (temperature == null) {
            return null;
        }
        return unit.getShortTemperatureText(context, temperature);
    }

    @Nullable
    public static String getTrendTemperature(Context context,
                                             @Nullable Integer night, @Nullable Integer day,
                                             TemperatureUnit unit) {
        if (night == null || day == null) {
            return null;
        }
        if (SettingsOptionManager.getInstance(context).isExchangeDayNightTempEnabled()) {
            return getShortTemperature(context, day, unit) + "/" + getShortTemperature(context, night, unit);
        } else {
            return getShortTemperature(context, night, unit) + "/" + getShortTemperature(context, day, unit);
        }
    }

    public boolean isValid() {
        return realFeelTemperature != null
                || realFeelShaderTemperature != null
                || apparentTemperature != null
                || windChillTemperature != null
                || wetBulbTemperature != null
                || degreeDayTemperature != null;
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/basic/models/weather/Temperature.kt
package wangdaye.com.geometricweather.common.basic.models.weather

import android.content.Context
import wangdaye.com.geometricweather.common.basic.models.options.unit.TemperatureUnit
import wangdaye.com.geometricweather.settings.SettingsManager.Companion.getInstance
import java.io.Serializable

/**
 * Temperature.
 * default unit : [TemperatureUnit.C]
 */
class Temperature(
    val temperature: Int? = null,
    val realFeelTemperature: Int? = null,
    val realFeelShaderTemperature: Int? = null,
    val apparentTemperature: Int? = null,
    val windChillTemperature: Int? = null,
    val wetBulbTemperature: Int? = null,
    val degreeDayTemperature: Int? = null
) : Serializable {

    companion object {
        @JvmStatic
        fun getTemperature(context: Context, temperature: Int?, unit: TemperatureUnit): String? {
            return if (temperature == null) null else unit.getValueText(context, temperature)
        }

        @JvmStatic
        fun getShortTemperature(context: Context, temperature: Int?, unit: TemperatureUnit): String? {
            return if (temperature == null) null else unit.getShortValueText(context, temperature)
        }

        @JvmStatic
        fun getTrendTemperature(context: Context, night: Int?, day: Int?, unit: TemperatureUnit): String? {
            return getTrendTemperature(
                context, night, day, unit,
                getInstance(context).isExchangeDayNightTempEnabled
            )
        }

        @JvmStatic
        fun getTrendTemperature(context: Context, night: Int?, day: Int?, unit: TemperatureUnit, switchDayNight: Boolean): String? {
            if (night == null || day == null) {
                return null
            }
            return if (switchDayNight) {
                getShortTemperature(context, day, unit) + "/" +
                        getShortTemperature(context, night, unit)
            } else {
                getShortTemperature(context, night, unit) + "/" +
                        getShortTemperature(context, day, unit)
            }
        }
    }

    val feelsLikeTemperature: Int? = realFeelTemperature ?: realFeelShaderTemperature ?: apparentTemperature
        ?: windChillTemperature ?: wetBulbTemperature ?: degreeDayTemperature

    fun getTemperature(context: Context, unit: TemperatureUnit): String? {
        return getTemperature(context, temperature, unit)
    }

    fun getShortTemperature(context: Context, unit: TemperatureUnit): String? {
        return getShortTemperature(context, temperature, unit)
    }

    fun getFeelsLikeTemperature(context: Context, unit: TemperatureUnit): String? {
        return getTemperature(context, feelsLikeTemperature, unit)
    }

    fun getShortFeelsLikeTemperature(context: Context, unit: TemperatureUnit): String? {
        return getShortTemperature(context, feelsLikeTemperature, unit)
    }
}
